### ISSUE TITLE
computeStyleTests() throws in Firefox because window.getComputedStyle(div) returns null

### ISSUE THREAD

**Issue by tedald** [jquery/jquery#3514]
### Description ###
jquery 3.1.1: Seeing occasional errors on our production site due to Firefox returning null for window.getComputedStyle(div) inside jquery support function computeStyleTests(). Related Firefox bug: https://bugzilla.mozilla.org/show_bug.cgi?id=548397

workaround in computeStyleTests: var divStyle = window.getComputedStyle( div ) || {}; 

### Link to test case ###
No test case yet. Perhaps as the Firefox bug states, iframes are involved.

**Comment by gibson042** [comment]
We can prevent the exception, but our CSS methods won't actually work without gCS. Further, `computeStyleTests` is called on-demand, and just loading jQuery doesn't constitute "demand". So what is happening that triggers the computation in a hidden iframe, and how much of a "fix" would a change be that prevents the exception but yields incorrect behavior?

**Comment by tedald** [comment]
> what is happening that triggers the computation in a hidden iframe
We are getting the errors via an error reporting service. It doesn't appear that the problem pages are actually in an iframe, but I don't yet have enough info.

> how much of a "fix" would a change be that prevents the exception but yields incorrect behavior?
An environment that returns null for getComputedStyle may indeed have other issues. But throwing an exception stops javascript block execution. I think fixing and limiting the issue to something more specific like "jquery css methods fail", while perhaps not always better, is generally preferable.

**Comment by dmethvin** [comment]
> We are getting the errors via an error reporting service. It doesn't appear that the problem pages are actually in an iframe, but I don't yet have enough info.

In that case this is definitely not something we should be sweeping under the rug just to keep the error log clean. If you determine that this particular error is harmless for your case, you can always filter it out of your error reporting.

**Comment by markelog** [comment]
Sounds like Gecko people not gonna fix that, moreover CSS specification might change too – https://github.com/w3c/csswg-drafts/issues/571

**Comment by timmywil** [comment]
FF's insistence not to fix is an issue: https://github.com/hakimel/reveal.js/issues/1546

**Comment by timmywil** [comment]
We'd like FF to address this. Dave to comment on the FF ticket.

**Comment by bzbarsky** [comment]
> Sounds like Gecko people not gonna fix that

Fix what?  The "getComputedStyle throws exceptions or returns null in display:none iframe" thing we _are_ aiming to fix.  https://bugzilla.mozilla.org/show_bug.cgi?id=548397 tracks that, as noted, and I have been working on it (albeit with limited success so far).  Unfortunately, it's a big architectural change and I've run into a _huge_ number of test failures with my attempted fixes, so I can't give a timeframe for it being fixed at the moment.  :(

The csswg-drafts discussion linked above would not affect getComputedStyle except for the small set of properties listed as returning used styles at <https://drafts.csswg.org/cssom/#resolved-value> (for which the used value would become equal to the computed value if box construction in the subframe were suppressed).  That said, computeStyleTests does in fact check some of those properties, so may be affected.

**Comment by markelog** [comment]
Some bits are confusing here, like

> would not affect getComputedStyle

vs

> so may be affected.

So sounds like we will be affected :)

> Fix what? The "getComputedStyle throws exceptions or returns null in display:none iframe" thing we are aiming to fix.

Iframe "situation" in FF was always problematic from my experience, remembering stuff like that – https://bugzilla.mozilla.org/show_bug.cgi?id=797029, so when you say

>  Unfortunately it's failures in all sorts of different parts of the code, not just a case of the same failure happening in lots of test suites.

I presumed situation is the similar (with some degree of doubt though, hence the "Sounds like" from original comment), especially considering that FF ticket was filled seven years ago.

I think I understand what you saying though, right now it seems we need clarify what to do or don't do on our side, should we wait for FF resolution and clarify exactly how css spec changes could affect us. 

Is that about right?

**Comment by bzbarsky** [comment]
> So sounds like we will be affected :)

Yes, if you actually rely on _layout_ in display:none iframes.  But that's really unrelated to the Gecko bug being cited, which breaks attempts to get _style_ inside display:none iframes.

**Comment by markelog** [comment]
> if you actually rely on layout in display:none iframes

That would happen if jQuery itself loaded in such iframe, right?

**Comment by bzbarsky** [comment]
Yes, indeed.

**Comment by timmywil** [comment]
I think we're going to count on a fix from FF here. Our support tests are pointless otherwise. Thank you @bzbarsky. Keep at it!

**Comment by bzbarsky** [comment]
You _could_ assume that exception means "supported".  Better than just completely failing....

**Comment by tedald** [comment]
Yes, or re-write the test to defer until successful. Completely failing breaks the script block.

**Comment by timmywil** [comment]
I'm not a fan of assuming support when the method doesn't even exist. However, I'm fine with deferring.

**Comment by xinqidianjiaoyu** [comment]
skladjf

**Comment by timmywil** [comment]
I started creating a test for this and it seems it's been fixed in FF?

**Comment by mgol** [comment]
@timmywil I removed the milestone & added the "Works For Me" label so that it doesn't get pulled into the changelog.
